src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2018 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// Custom CFG serialiser for Yorick.
+/// At the time of writing no crate using `proc_macro` can be used in-compiler, otherwise we'd have
+/// used Serde.
+
+use rustc::ty::TyCtxt;
+
+use rustc::hir::def_id::DefId;
+use rustc::mir::{Mir, TerminatorKind, Operand, Constant, BasicBlock};
+use rustc::ty::{TyS, TyKind, Const};
+use rustc::util::nodemap::DefIdSet;
+use std::path::PathBuf;
+use std::fs::File;
+use rustc_yk_link::YkExtraLinkObject;
+use std::fs;
+use byteorder::{NativeEndian, WriteBytesExt};
+
+// Edge kinds.
+const GOTO: u8 = 0;
+const SWITCHINT: u8 = 1;
+const RESUME: u8 = 2;
+const ABORT: u8 = 3;
+const RETURN: u8 = 4;
+const UNREACHABLE: u8 = 5;
+const DROP_NO_UNWIND: u8 = 6;
+const DROP_WITH_UNWIND: u8 = 7;
+const DROP_AND_REPLACE_NO_UNWIND: u8 = 8;
+const DROP_AND_REPLACE_WITH_UNWIND: u8 = 9;
+const CALL_NO_CLEANUP: u8 = 10;
+const CALL_WITH_CLEANUP: u8 = 11;
+const CALL_UNKNOWN_NO_CLEANUP: u8 = 12;
+const CALL_UNKNOWN_WITH_CLEANUP: u8 = 13;
+const ASSERT_NO_CLEANUP: u8 = 14;
+const ASSERT_WITH_CLEANUP: u8 = 15;
+const YIELD_NO_DROP: u8 = 16;
+const YIELD_WITH_DROP: u8 = 17;
+const GENERATOR_DROP: u8 = 18;
+const FALSE_EDGES: u8 = 19;
+const FALSE_UNWIND: u8 = 20;
+const NO_MIR: u8 = 254;
+const SENTINAL: u8 = 255;
+
+const MIR_CFG_SECTION_NAME: &'static str = ".yk_mir_cfg";
+const SECTION_VERSION: u16 = 0;
+
+/// Serialises the control flow for the given `DefId`s into a ELF object file and returns a handle
+/// for linking.
+pub fn emit_mir_cfg_section<'a, 'tcx, 'gcx>(
+    tcx: &'a TyCtxt<'a, 'tcx, 'gcx>, def_ids: &DefIdSet, exe_filename: PathBuf)
+    -> YkExtraLinkObject {
+
+    // Serialise the MIR into a file whose name is derived from the output binary. The filename
```

Comment:
This fixes the reproducible build tests. We had used a tempfile with a random name (via `mkstemp`) before.